src/marqo/core/index_management/vespa_application_package.py

```diff
@@ -0,0 +1,644 @@
+    _MARQO_INDEX_SETTINGS_HISTORY_FILE = 'marqo_index_settings_history.json'
+    _BACKUP_FILE = 'app_bak.tgz'
+    _COMPONENTS_JAR_FOLDER = Path(__file__).parent / '../../../../vespa/target'
+
+    def __init__(self, store: VespaApplicationStore):
+        self._store = store
+        self.is_configured = self._store.file_exists(self._MARQO_CONFIG_FILE)
+        self._service_xml = ServiceXml(self._store.read_text_file(self._SERVICES_XML_FILE))
+        self._marqo_config_store = MarqoConfigStore(self._store.read_text_file(self._MARQO_CONFIG_FILE))
+        self._index_setting_store = IndexSettingStore(
+            self._store.read_text_file(self._MARQO_INDEX_SETTINGS_FILE) or '{}',
+            self._store.read_text_file(self._MARQO_INDEX_SETTINGS_HISTORY_FILE) or '{}')
+
+    def get_marqo_config(self) -> MarqoConfig:
+        return self._marqo_config_store.get()
+
+    def need_bootstrapping(self, marqo_version: str, marqo_config_doc: Optional[MarqoConfig] = None) -> bool:
+        """
+        Bootstrapping is needed when
+        - the version of Marqo is higher than the version of the deployed application
+        - the version of Marqo is lower but allow downgrade is enabled (used for rollback)
+        The version of the application is retrieved in the following order:
+        - from the 'marqo_config.json' file (Post v2.12.0)
+        - from the marqo_config_doc passed in which is marqo__config doc saved in marqo__settings schema (Post v2.1.0)
+        - if neither is available, return 2.0.0 as default (Pre v2.1.0 or not bootstrapped yet)
+        """
+        logger.info(f'Marqo version is {marqo_version}')
+
+        if self.is_configured and self.get_marqo_config() is not None:
+            app_version = self.get_marqo_config().version
+            logger.info(f'Vespa app version is detected in marqo_config.json: {app_version}')
+        elif marqo_config_doc:
+            app_version = marqo_config_doc.version
+            logger.info(f'Vespa app version is detected in marqo__config doc: {app_version}')
+        else:
+            app_version = '2.0.0'
+            logger.info(f'Vespa app version is not detected, set to default: {app_version}')
+
+        marqo_sem_version = semver.VersionInfo.parse(marqo_version, optional_minor_and_patch=True)
+        app_sem_version = semver.VersionInfo.parse(app_version, optional_minor_and_patch=True)
+
+        return app_sem_version < marqo_sem_version
+
+    def bootstrap(self, marqo_version: str, existing_index_settings: List[MarqoIndex] = ()) -> None:
+        logger.info(f'Bootstrapping Vespa app to {marqo_version}')
+        # Migrate existing index settings from previous versions of Marqo
+        if not self.is_configured:
+            if existing_index_settings is not None and len(existing_index_settings) > 0:
+                logger.info(f'Importing existing index settings {[index.json for index in existing_index_settings]}')
```

Comment:
@farshidz, Do we have any security concerns of logging out the existing index settings?